Panel-based Datashader dashboard

jbednar opened the pull request:
This PR closes #358, #618, and #221 by providing a [Panel](http://panel.pyviz.org)-based dashboard to replace the current pure Bokeh version that has been broken for several releases now, with the Intake support provided by @jsignell.  The old Bokeh dashboard was over 500 lines of code that I was never able to update or even keep running, hence the long period of breakage.  This new version fits into 50 lines of code (so far) and has nearly the same functionality, once the to-do items below are addressed:

- [x] opacity (of the data) and spread_size are not hooked up to the control
- [x] the widgets aren't in any order yet
- [ ] it would be nice to separate some widgets to be on top of the plot and others down the side, if there is a clear semantic distinction
- [x] colormaps should be updated when the plot changes
- [ ] opensky data is not tested
- [x] osm data is not tested and is not yet listed in the .yml file
- [ ] the basemap argument doesn't update the map tiles, due to an open Bokeh bug in caching tiles
- [x] should probably add a logo
- [x] Need to allow field (fare, etc.) to be selected

There are also other to-do items that weren't supported in the last working version of the dashboard:

- [ ] Hover info (#252)
- [ ] Colorbar and/or categorical legend

philippjfr commented on examples/dashboard.yml:
```diff
@@ -0,0 +1,100 @@
+    description: Large version of nyc taxi dataset
+    driver: parquet
+    args:
+      urlpath: '{{ CATLOG_DIR }}data/nyc_taxi_wide.parq'
```
Bit confused here, ``nyc_taxi_wide.parq`` is still not listed in the datashader datasets.yaml, it is only in the pyviz on.

jbednar replied:
Good point. I think we should probably delete the .csv version; .csv is really not a recommended format for Datashader.

philippjfr commented on examples/dashboard.yml:
```diff
@@ -0,0 +1,100 @@
+        counts:
+          label: Ride counts
+      plot:
+        xlim: !!python/tuple [-8240227.037, -8231283.905]
```
I think I've arranged it so you can simply provide lists and they are cast appropriately.

philippjfr replied:
Actually I'm probably wrong about that, I think that was explicitly for dimension ranges.

jbednar replied:
Would be good to provide support for such casting if you can, to avoid this awkwardness.

jbednar replied:
Oh, it's definitely YAML's fault, or at least JSON's fault (from which YAML inherits its scope).  But the fact remains that it's awkward for Python users to write code like the above, so we should probably accept a list as well as a tuple unless we are already using the list/tuple distinction for some semantic purpose in this case.

jbednar replied:
Fixed in https://github.com/pyviz/hvplot/pull/248.

philippjfr commented on examples/dashboard.yml:
```diff
@@ -0,0 +1,100 @@
+    description: OpenSky Flight Paths
+    driver: netcdf
+    args:
+      urlpath: '{{ CATLOG_DIR }}data/opensky.h5'
```
Is ``{{ CATLOG_DIR }}`` really the correct way to spell this? I've always used ``{{ CATALOG_DIR }}/``?

jbednar replied:
I think it looks more natural with the /, but it seems like both ways work?

jbednar replied:
Oh, you mean CAT LOG?  That seems like a typo.

jbednar replied:
This one hasn't been tested, so I'll just go ahead and fix that.

philippjfr replied:
I meant both things, glad it's a typo though.

philippjfr replied:
>This one hasn't been tested, so I'll just go ahead and fix that.

The typo appears in all of them.

jbednar replied:
Odd; not sure how it's ever worked, then. "CATLOG" doesn't appear in the intake docs or repo.

jbednar replied:
(Scratches head) Odd.  I've fixed that and pushed.

philippjfr replied:
The variable was probably just replaced with an empty string so it would work as long as the dashboard.ipynb is in the same place as the data directory.

philippjfr commented:
I've pushed updates to apply opacity and spread_size, however control over global opacity required a small fix in holoviews, which we can push out as part of a 1.11.1 release (see https://github.com/ioam/holoviews/pull/3371).

jbednar commented:
Current version:

![image](https://user-images.githubusercontent.com/1695496/51285575-fe229e00-19b4-11e9-9746-b2dcb2dff6ff.png)

Still just 54 lines.  Need to add support for selecting which field to plot.

jsignell commented on examples/dashboard.yml:
```diff
@@ -0,0 +1,120 @@
+    description: OpenSky Flight Paths
+    driver: netcdf
+    args:
+      urlpath: '{{ CATALOG_DIR }}/data/opensky.h5'
```
I think you need to add a chunks argument, but you can set it to empty:

```suggestion
      urlpath: '{{ CATALOG_DIR }}/data/opensky.h5'
      chunks: {}
```

jsignell commented on examples/dashboard.yml:
```diff
@@ -0,0 +1,121 @@
+      plot:
+        xlim: !!python/tuple [-8240227.037, -8231283.905]
+        ylim: !!python/tuple [4974203.152, 4979238.441]
+        kind: points
```
If you need to have access to other columns later, then we can include them in the original points plot

```suggestion
        kind: points
        hover_cols: ['dropoff_hour', 'pickup_hour', 'passenger_count']
```

jsignell commented on examples/dashboard.ipynb:
```diff
@@ -0,0 +1,139 @@
+    "\n",
+    "    def view(self,**kwargs):\n",
+    "        points   = hv.DynamicMap(self.points)\n",
+    "        agg      = rasterize(points, width=600, height=400)\n",
```
To aggregate on another field, you need to add an aggregator to the rasterize call. The field can be another param stream. 

```suggestion
    "        agg      = rasterize(points, width=600, height=400, aggregator=rd.mean('passenger_count'))\n",
```

jsignell commented on examples/dashboard.ipynb:
```diff
@@ -0,0 +1,139 @@
+    "norms = {'Histogram_Equalization': 'eq_hist', 'Linear': 'linear', 'Log': 'log', 'Cube root': 'cbrt'}\n",
+    "maps  = ['CartoMidnight', 'StamenWatercolor', 'StamenTonerBackground', 'EsriImagery', 'EsriUSATopo', 'EsriTerrain']\n",
+    "bases = {name: ts.relabel(name) for name, ts in gts.tile_sources.items() if name in maps}\n",
+    "plots = {source.metadata['plots'][p].get('label', p):p for p in source.plots}\n",
```
So just limit the list of fields to ones that you want to allow aggregation over and have one just be counts or something if you want to allow that. 

This will get a dict of labelled selectors for all the available aggregator fields. 

```suggestion
    "plots = {source.metadata['plots'][p].get('label', p):p for p in source.plots}\n",
    "fields = {k: v.get('label', k), for k, v in source.metadata['fields'].items()}\n",
```

jbednar commented:
I've made a number of changes in the above commits:

- Added `nyc_taxi_50k` to datasets.yml as a quick test case to use
- Updated the .yml file to consistently list "counts" as the first item in the list of plottable fields, as it makes a good default.  
- Removed any mention of "fare" from the .yml, because the nyc_taxi parq files don't currently include the fare column.  (At some point I need to make new Parquet files including fare and also adding JFK to the visible range of points.)
- Added an "agg_fn" parameter that allows the user to select min, max, mean, etc.

All of that seems fine so far. I've also made some changes that don't quite work yet:

- [ ] I changed the `.points` method to `.rasterize` so that it could specify the agg_fn, which works but required me to make `rasterize` no longer be dynamic, which means that RangeXY is no longer tied to this object and zooming no longer updates the datashaded plot.
- [ ] I added a "field" parameter, which is populated correctly but doesn't seem to have any effect yet, despite being applied in the same method as the agg_fn and having the same dependency structure.

jbednar commented:
Thanks, @philippjfr for re-connecting RangeXY and for realizing that the field parameter _is_ working, it just has no noticeable effect for a counts aggregator; e.g. switch to `Mean` to see the field having an effect. I've renamed `.points` to `.elem` because the plot isn't always a points plot; depends what's declared in the data catalog.  

I also reorganized the code a bit and added a very extensive section at the end unpacking how it works.  It would be nice not to have to use three different mechanisms for getting the DynamicMaps involved, but I'll open a separate issue about that.  Meanwhile, I think it's a very concise and *mostly* understandable way to express some quite complicated relationships between user-visible controls and computations, and it's a massive improvement over the old dashboard, so it should be ready to merge when tests pass.
